index.js

```diff
@@ -240,36 +240,40 @@ FSWatcher.prototype._awaitWriteFinish = function(path, threshold, callback) {
   (function awaitWriteFinish (prevStat) {
     fs.stat(fullPath, function(err, curStat) {
       if (err) {
-        delete this._pendingWrites[path];
-        if (err.code == 'ENOENT') return;
-        return callback(err);
+        // prevent `unlink` race condition
+        this._getWatchedDir(sysPath.dirname(path))
+          .remove(sysPath.basename(path));
+        if (this._pendingWrites[path]) this._pendingWrites[path].cancelWait();
+        if (err.code !== 'ENOENT') callback(err);
+        return;
```

Comment:
The callback doesn't necessarily need to be called to do what it currently does. It only emits either an error or the event, and sometimes we want neither. I agree that it's ugly though, and as a next step I wanted to look into refactoring further to deal with cleaning up this control flow. 

Please open your version in a PR so we can see how it fares on appveyor, then we can determine which direction seems better.